File: Miscellaneous/Golang/543. Diameter of Binary Tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{1, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Left.Left = &TreeNode{4, nil, nil}
	root.Left.Right = &TreeNode{5, nil, nil}
	root.Right = &TreeNode{3, nil, nil}
	result := diameterOfTree(root)
	fmt.Println(result)

}

func height(root *TreeNode) int {
	if root == nil {
		return 0
	}

	return 1 + max(height(root.Left), height(root.Right))
}

func diameterOfTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	lheight := height(root.Left)
	rheight := height(root.Right)

	ldiameter := diameterOfTree(root.Left)
	rdiameter := diameterOfTree(root.Right)

	return max(lheight+rheight, max(ldiameter, rdiameter))

}

func max(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: Miscellaneous/Python/543. Diameter of Binary Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def height(root):
    if root is None:
        return 0
    
    return 1 + max(height(root.left), height(root.right))

class Solution:
    def diameterOfBinaryTree(self, root):
        if root is None:
            return 0
        
        lheight = height(root.left)
        rheight = height(root.right)

        ldiameter = self.diameterOfBinaryTree(root.left)
        rdiameter = self.diameterOfBinaryTree(root.right)
        
        return max(lheight + rheight, max(ldiameter, rdiameter))

obj = Solution()
root = TreeNode(4)
root.left = TreeNode(2)
root.left.left = TreeNode(1)
root.left.right = TreeNode(3)
root.right = TreeNode(7)
root.right.left = TreeNode(6)
root.right.right = TreeNode(9)
ans = obj.diameterOfBinaryTree(root)
print(ans)